File: platforms/android/stato/src/main/java/com/facebook/stato/plugins/common/MainThreadStatoReceiver.java
/*
 *  Copyright (c) 2018-present, Facebook, Inc.
 *
 *  This source code is licensed under the MIT license found in the LICENSE
 *  file in the root directory of this source tree.
 *
 */
package com.facebook.stato.plugins.common;

import android.os.Handler;
import android.os.Looper;
import com.facebook.stato.core.StatoObject;
import com.facebook.stato.core.StatoReceiver;
import com.facebook.stato.core.StatoResponder;
import java.io.PrintWriter;
import java.io.StringWriter;

public abstract class MainThreadStatoReceiver implements StatoReceiver {

  private final Handler mHandler = new Handler(Looper.getMainLooper());

  @Override
  public final void onReceive(final StatoObject params, final StatoResponder responder) {
    mHandler.post(
        new Runnable() {
          @Override
          public void run() {
            try {
              onReceiveOnMainThread(params, responder);
            } catch (Exception ex) {
              responder.error(
                  new StatoObject.Builder()
                      .put("name", ex.getClass().getCanonicalName())
                      .put("message", ex.getMessage())
                      .put("stacktrace", getStackTraceString(ex))
                      .build());
            }
          }
        });
  }

  private static String getStackTraceString(Throwable th) {
    StringWriter stringWriter = new StringWriter();
    th.printStackTrace(new PrintWriter(stringWriter));
    return stringWriter.toString();
  }

  public abstract void onReceiveOnMainThread(StatoObject params, StatoResponder responder)
      throws Exception;
}

File: platforms/android/android-stato/src/main/kotlin/org/stato/plugins/common/MainThreadStatoReceiver.kt
/*
 *  Copyright (c) 2018-present, Facebook, Inc.
 *
 *  This source code is licensed under the MIT license found in the LICENSE
 *  file in the root directory of this source tree.
 *
 */
package org.stato.plugins.common

import android.os.Handler
import android.os.Looper
import org.stato.core.StatoObject
import org.stato.core.StatoReceiver
import org.stato.core.StatoReceiverCallback
import org.stato.core.StatoResponder
import java.io.PrintWriter
import java.io.StringWriter

abstract class MainThreadStatoReceiver : StatoReceiverCallback {

  private val mHandler = Handler(Looper.getMainLooper())

  @Throws(Exception::class)
  abstract fun onReceiveOnMainThread(params: StatoObject, responder: StatoResponder)

  //override fun onReceive(params: StatoObject, responder: StatoResponder) {
  override fun invoke(params: StatoObject, responder: StatoResponder) {
    mHandler.post {
      try {
        onReceiveOnMainThread(params, responder)
      } catch (ex: Exception) {
        responder.error(
          StatoObject.Builder()
            .put("name", ex.javaClass.canonicalName!!)
            .put("message", ex.message)
            .put("stacktrace", getStackTraceString(ex))
            .build())
      }

    }
  }


  private fun getStackTraceString(th: Throwable): String {
    return StringWriter().apply {
      th.printStackTrace(PrintWriter(this))
    }.toString()
  }


}
